Source/MBF.SAMCD.ADC.pas

```diff
@@ -0,0 +1,225 @@
+  License for more details.
+}
+{< Atmel SAMD series ADC functions. }
+
+interface
+
+{$include MBF.Config.inc}
+
+uses
+  MBF.SAMCD.Helpers,
+  MBF.SAMCD.GPIO;
+
+const
+{$if defined(samd10) or defined(samd21)}
+  ADCMux = PORT_PMUX_B_Val;
+{$elseif defined(samc21)}
+  ADCMux = PORT_PMUX_B_Val;
+  ADC1Mux = PORT_PMUX_B_Val;
+{$else}
+  {$error Unknown Chip, please check mbf.boards.samd10.inc and then define ADC Mux here}
+{$endif}
+
+type
+  TADCHelper = record helper for TADC_Registers
+  private
+    procedure SyncWait;
+    procedure ConversionWait;
+    function GetRawValue(const Channel: TPinIdentifier): Cardinal;
+  public
+    procedure Initialize;
+    property GetADCResult[const Channel: TPinIdentifier]:Cardinal read GetRawValue;
+  end;
+
+implementation
+
+uses
+  MBF.BitHelpers,
+  MBF.SAMCD.SystemCore;
+
+type
+  {$ifdef samd10}
+  TADCPin = (AIN0,AIN1,AIN2,AIN3,AIN4,AIN5,AIN6,AIN7,AIN8,AIN9,None);
+  {$endif}
+  {$ifdef samd21}
+  TADCPin = (AIN0,AIN1,AIN2,AIN3,AIN4,AIN5,AIN6,AIN7,AIN8,AIN9,AIN10,AIN11,AIN12,AIN13,AIN14,AIN15,AIN16,AIN17,AIN18,AIN19,None);
+  {$endif}
+  {$ifdef samc21}
+  TADCPin = (AIN0,AIN1,AIN2,AIN3,AIN4,AIN5,AIN6,AIN7,AIN8,AIN9,AIN10,AIN11,None);
+  TADC1Pin = (AIN0,AIN1,AIN2,AIN3,AIN4,AIN5,AIN6,AIN7,AIN8,AIN9,AIN10,AIN11,None);
+  {$endif}
+
+  TADCPins = set of TADCPin;
+  {$ifdef samc21}
+  TADC1Pins = set of TADC1Pin;
+  {$endif}
+
+const
+  {$ifdef samd10}
+  TADCPinsMap : array[TADCPin] of TPinIdentifier=(
+      TNativePin.PA2,TNativePin.PA3,TNativePin.PA4,TNativePin.PA5,TNativePin.PA6,TNativePin.PA7,TNativePin.PA14,TNativePin.PA15,
+      TNativePin.PA10,TNativePin.PA11,TNativePin.None);
+  {$endif}
+
+  {$ifdef samd21}
+  TADCPinsMap : array[TADCPin] of TPinIdentifier=(
+      TNativePin.PA2,TNativePin.PA3,TNativePin.PB8,TNativePin.PB9,TNativePin.PA4,TNativePin.PA5,TNativePin.PA6,TNativePin.PA7,
+      TNativePin.PB0,TNativePin.PB1,TNativePin.PB2,TNativePin.PB3,TNativePin.PB4,TNativePin.PB5,TNativePin.PB6,TNativePin.PB7,
+      TNativePin.PA8,TNativePin.PA9,TNativePin.PA10,TNativePin.PA11,TNativePin.None);
+  {$endif}
+
+  {$ifdef samc21}
+  TADCPinsMap : array[TADCPin] of TPinIdentifier=(
+      TNativePin.PA2,TNativePin.PA3,TNativePin.PB8,TNativePin.PB9,TNativePin.PA4,TNativePin.PA5,TNativePin.PA6,TNativePin.PA7,
+      TNativePin.PA8,TNativePin.PA9,TNativePin.PA10,TNativePin.PA11,TNativePin.None);
+  TADC1PinsMap : array[TADC1Pin] of TPinIdentifier=(
+      TNativePin.PB0,TNativePin.PB1,TNativePin.PB2,TNativePin.PB3,TNativePin.PB8,TNativePin.PB9,TNativePin.PB4,TNativePin.PB5,
+      TNativePin.PB6,TNativePin.PB7,TNativePin.PA8,TNativePin.PA9,TNativePin.None);
+  {$endif}
+
+procedure TADCHelper.SyncWait;
+begin
+  {$ifdef samd}
+  while (GetBit(Self.STATUS,ADC_STATUS_SYNCBUSY_Pos)) do begin end;   // Wait for synchronization
+  {$endif}
+  {$ifdef samc}
+  while (Self.SYNCBUSY>0) do begin end;   // Wait for synchronization
+  {$endif}
+end;
```

Comment:
You should throw an error when no definition matches, this makes extending mbf a lot easier

Reply:
For sure !!